### ISSUE TITLE
contract between symmetric and nonsymmetric rank2 tensors

### ISSUE THREAD

**Issue by davydden** [dealii/dealii#6374]
in CM one needs single contraction operation `T * S` between non-symmetric `T` and symmetric `S` rank 2 tensors. Currently to implement tangents one has to copy a symmetric tensor into non-symmetric one, see [code gallery example](https://github.com/dealii/code-gallery/blob/master/Quasi_static_Finite_strain_Compressible_Elasticity/cook_membrane.cc#L1751), where the term where this is done can be also written as `T*S`.

@jppelteret what do you think? 

@tamiko i think such operation won't fit into your templated `contract` or `double_contract`. So it probably just has to be a stand-alone function like those which are deprecated now.

**Comment by jppelteret** [comment]
It looks like [this function is implemented](https://dealii.org/developer/doxygen/deal.II/classSymmetricTensor.html#a075d25d60b04456f2d116bfa6dcc4a09) and has been for a long time (right?). I have no clue as to why I would not / could not have used this function in this context. I'll try later on to see if it works in the code gallery examples.

**Comment by davydden** [comment]
>  this function is implemented and has been for a long time

yes, that would work for your code-gallery example, but if you remove ` if (component_i == component_j)` then the same result is given by `T * S`, and this one AFAIK is not implemented.

**Comment by jppelteret** [comment]
Oh, you're talking about `T` being a rank-2 tensor? I didn't get that.

**Comment by jppelteret** [comment]
Yeah, I'm not sure where my head was when I first replied. Sorry about that :-/
